robots/pure_pursuit_controller.h

```diff
@@ -0,0 +1,169 @@
+                const auto x2 = (D * dy - sgn(dy.value()) * dx * sqrt(discriminant)) / (d*d);
+                // y components of the intersection point
+                const auto y1 = (-D * dx + fabs(dy) * sqrt(discriminant)) / (d*d);
+                const auto y2 = (-D * dx - fabs(dy) * sqrt(discriminant)) / (d*d);
+
+                // if there is 2 intersections possible with the circle, we select the second one, 
+                // as that will be closer to the end point of the segment
+                const bool validIntersection1 = (min(p1x, p2x) < x1 && x1 < max(p1x, p2x)) ||
+                                          (min(p1y, p2y) < y1 && y1 < max(p1y,p2y));
+                const bool validIntersection2 = (min(p1x, p2x) < x2 && x2 < max(p1x, p2x)) ||
+                                          (min(p1y, p2y) < y2 && y2 < max(p1y,p2y));
+
+                // we always want the latest path segment point so if we have a 
+                // valid point, we delete the previous point
+                if (validIntersection1) {
+                    lookaheadVector.x() = x1 + x;
+                    lookaheadVector.y() = y1 + y;
+                }
+
+                // if there is a valid 2. intersection point, we keep that and remove
+                // the first one
+                if (validIntersection2) {
+                    if (lookaheadVector.isnan() || fabs(x1 - p2x) > fabs(x2 - p2x) ||
+                        fabs(y1 - p2y) > fabs(y2 - p2y)) {
+                            
+                        lookaheadVector.x() = x2 + x;
+                        lookaheadVector.y() = y2 + y;
+                    }
+                }
+            }
+        } 
+        return lookaheadVector;
+    }
+
+private:
+
+    millimeter_t m_lookAheadDistance;                 // the distance to look ahead
+    millimeter_t m_wheelBase;                         // length between wheel bases
+    std::vector<Vector2<millimeter_t>> m_wayPoints;   // list of waypoint coordinates
+
+    // computes the turning angle using the lookahead point
+    degree_t computeTurningAngle(const millimeter_t xrobot, const millimeter_t yrobot 
+                                ,const millimeter_t xlookahead, const millimeter_t ylookahead
+                                ,const radian_t heading) {
+        using namespace units::math;
+        // calculating bearing [robot-lookahead point]
+        const auto dx = xlookahead-xrobot;
+        const auto dy = ylookahead-yrobot;
+        const degree_t bearing = atan2(dy,dx)-heading;
+       
+        // calculating turning angle
+        const auto xlength = sin(bearing) * m_lookAheadDistance;
+
+        // calculating arc to turn
+        const auto kb = (2*xlength*m_wheelBase)/(pow<2>(m_lookAheadDistance));
+        const degree_t turningAngle = atan(kb);
+        return turningAngle;
+    }
+
+    // sign function
+    template <typename T> int sgn(T val) {
```

Comment:
This should be ``static`` because it doesn't use any of the member vars/methods. Could also be marked ``constexpr``